### PR TITLE
refactor(payment_attempt_v2): add payment attempt v2 domain and diesel models

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] Refactoring

## Description

This PR adds new v2 structs for payment attempt diesel and domain models. This PR also includes changes to put most of the existing payments code under the v1 feature flag. 

The intermediate feature flag `payment_v2` has been removed

## How did you test it?

Normal sanity testing should ensure that there are not breaking changes in this PR.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`

### CODE REVIEW

File: crates/diesel_models/src/payment_attempt.rs [hunk 1]
```diff
@@ -4,35 +4,24 @@ use serde::{Deserialize, Serialize};
...
     pub status: storage_enums::AttemptStatus,
-    pub amount: MinorUnit,
-    pub currency: Option<storage_enums::Currency>,
-    pub save_to_locker: Option<bool>,
     pub connector: Option<String>,
     pub error_message: Option<String>,
-    pub offer_amount: Option<MinorUnit>,
     pub surcharge_amount: Option<MinorUnit>,
-    pub tax_amount: Option<MinorUnit>,
     pub payment_method_id: Option<String>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:24]: If we have `GlobalPaymentMethodId` available, we can change `String` to the `GlobalPaymentMethodId`.

File: crates/diesel_models/src/payment_attempt.rs [hunk 2]
```diff
@@ -4,35 +4,24 @@ use serde::{Deserialize, Serialize};
...
-    pub payment_method: Option<storage_enums::PaymentMethod>,
-    pub connector_transaction_id: Option<String>,
-    pub capture_method: Option<storage_enums::CaptureMethod>,
-    #[serde(default, with = "common_utils::custom_serde::iso8601::option")]
-    pub capture_on: Option<PrimitiveDateTime>,
     pub confirm: bool,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:25]: Is this field required?

File: crates/diesel_models/src/payment_attempt.rs [hunk 3]
```diff
@@ -81,11 +62,19 @@ pub struct PaymentAttempt {
...
     pub organization_id: id_type::OrganizationId,
     pub card_network: Option<String>,
+    pub payment_method_type_v2: Option<storage_enums::PaymentMethod>,
+    pub connector_payment_id: Option<String>,
+    pub payment_method_subtype: Option<storage_enums::PaymentMethodType>,
+    pub routing_result: Option<serde_json::Value>,
+    pub authentication_applied: Option<common_enums::AuthenticationType>,
+    pub external_reference_id: Option<String>,
+    pub tax_on_surcharge: Option<MinorUnit>,
+    pub id: String,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:72]: Use GlobalId type instead if String

File: crates/diesel_models/src/payment_attempt.rs [hunk 4]
```diff
@@ -81,11 +62,19 @@ pub struct PaymentAttempt {
     pub profile_id: id_type::ProfileId,
     pub organization_id: id_type::OrganizationId,
     pub card_network: Option<String>,
+    pub payment_method_type_v2: Option<storage_enums::PaymentMethod>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:65]: Change the Enums too for `PaymentMethod` and `PaymentMethodType` to `PaymentMethodTypeV2` and `PaymentMethodSubType`.

File: crates/diesel_models/src/payment_attempt.rs [hunk 5]
```diff
@@ -81,11 +62,19 @@ pub struct PaymentAttempt {
     pub profile_id: id_type::ProfileId,
     pub organization_id: id_type::OrganizationId,
     pub card_network: Option<String>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:64]: We should not directly flatten the payment_method data in attempt, remove this field for now.

File: crates/diesel_models/src/payment_attempt.rs [hunk 6]
```diff
@@ -43,22 +32,15 @@ pub struct PaymentAttempt {
     pub last_synced: Option<PrimitiveDateTime>,
     pub cancellation_reason: Option<String>,
     pub amount_to_capture: Option<MinorUnit>,
-    pub mandate_id: Option<String>,
     pub browser_info: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:35]: Make all serde_json::Value as Secret

File: crates/diesel_models/src/payment_attempt.rs [hunk 7]
```diff
@@ -71,7 +53,6 @@ pub struct PaymentAttempt {
     pub external_three_ds_authentication_attempted: Option<bool>,
     pub authentication_connector: Option<String>,
     pub authentication_id: Option<String>,
-    pub mandate_data: Option<storage_enums::MandateDetails>,
     pub fingerprint_id: Option<String>,
     pub payment_method_billing_address_id: Option<String>,
     pub charge_id: Option<String>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:58]: Remove this for now, we add these later if required

File: crates/diesel_models/src/payment_attempt.rs [hunk 8]
```diff
@@ -71,7 +53,6 @@ pub struct PaymentAttempt {
     pub external_three_ds_authentication_attempted: Option<bool>,
     pub authentication_connector: Option<String>,
     pub authentication_id: Option<String>,
-    pub mandate_data: Option<storage_enums::MandateDetails>,
     pub fingerprint_id: Option<String>,
     pub payment_method_billing_address_id: Option<String>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:57]: This field is not required

File: crates/diesel_models/src/payment_attempt.rs [hunk 9]
```diff
@@ -71,7 +53,6 @@ pub struct PaymentAttempt {
     pub external_three_ds_authentication_attempted: Option<bool>,
     pub authentication_connector: Option<String>,
     pub authentication_id: Option<String>,
-    pub mandate_data: Option<storage_enums::MandateDetails>,
     pub fingerprint_id: Option<String>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:56]: Is this required? Is this a deprecated field?

File: crates/diesel_models/src/payment_attempt.rs [hunk 10]
```diff
@@ -43,22 +32,15 @@ pub struct PaymentAttempt {
...
-    pub straight_through_algorithm: Option<serde_json::Value>,
     pub preprocessing_step_id: Option<String>,
-    // providing a location to store mandate details intermediately for transaction
-    pub mandate_details: Option<storage_enums::MandateDataType>,
     pub error_reason: Option<String>,
     pub multiple_capture_count: Option<i16>,
-    // reference to the payment at connector side
     pub connector_response_reference_id: Option<String>,
     pub amount_capturable: MinorUnit,
     pub updated_by: String,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:46]: We have a field called encoded_data, what is the purpose of it? If that is required make it secret or encrypted.

Comment (jarnura) [reply crates/diesel_models/src/payment_attempt.rs:46]: Add connector_id and avoid using merchant_connector_id

File: crates/diesel_models/src/payment_attempt.rs [hunk 11]
```diff
@@ -81,11 +62,19 @@ pub struct PaymentAttempt {
     pub profile_id: id_type::ProfileId,
     pub organization_id: id_type::OrganizationId,
     pub card_network: Option<String>,
+    pub payment_method_type_v2: Option<storage_enums::PaymentMethod>,
+    pub connector_payment_id: Option<String>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:66]: What is difference bewteen connector_payment_id and connector_response_reference_id?

File: crates/diesel_models/src/payment_attempt.rs [hunk 12]
```diff
@@ -43,22 +32,15 @@ pub struct PaymentAttempt {
...
     pub browser_info: Option<serde_json::Value>,
     pub error_code: Option<String>,
     pub payment_token: Option<String>,
     pub connector_metadata: Option<serde_json::Value>,
     pub payment_experience: Option<storage_enums::PaymentExperience>,
-    pub payment_method_type: Option<storage_enums::PaymentMethodType>,
     pub payment_method_data: Option<serde_json::Value>,
-    pub business_sub_label: Option<String>,
-    pub straight_through_algorithm: Option<serde_json::Value>,
     pub preprocessing_step_id: Option<String>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:41]: Can we make preprocessing_step_details instead having a single id.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/diesel_models/src/payment_attempt.rs [hunk 13]
```diff
@@ -81,11 +62,19 @@ pub struct PaymentAttempt {
...
+    pub payment_method_type_v2: Option<storage_enums::PaymentMethod>,
+    pub connector_payment_id: Option<String>,
+    pub payment_method_subtype: Option<storage_enums::PaymentMethodType>,
+    pub routing_result: Option<serde_json::Value>,
+    pub authentication_applied: Option<common_enums::AuthenticationType>,
+    pub external_reference_id: Option<String>,
+    pub tax_on_surcharge: Option<MinorUnit>,
+    pub id: String,
     pub shipping_cost: Option<MinorUnit>,
     pub order_tax_amount: Option<MinorUnit>,
```
Comment (jarnura) [inline crates/diesel_models/src/payment_attempt.rs:74]: Add version field also in Attempts

File: crates/hyperswitch_domain_models/src/payments/payment_attempt.rs [hunk 14]
```diff
@@ -162,6 +163,111 @@ pub trait PaymentAttemptInterface {
     ) -> error_stack::Result<i64, errors::StorageError>;
 }
 
+#[cfg(feature = "v2")]
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct PaymentAttempt {
+    pub payment_id: id_type::GlobalPaymentId,
+    pub merchant_id: id_type::MerchantId,
+    pub status: storage_enums::AttemptStatus,
+    pub net_amount: MinorUnit,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/payments/payment_attempt.rs:172]: Use `AmountBreakup` in Domain models, while doing remove field in plain structure.

File: crates/hyperswitch_domain_models/src/payments/payment_attempt.rs [hunk 15]
```diff
@@ -162,6 +163,111 @@ pub trait PaymentAttemptInterface {
...
+    pub confirm: bool,
+    pub authentication_type: Option<storage_enums::AuthenticationType>,
+    #[serde(with = "common_utils::custom_serde::iso8601")]
+    pub created_at: PrimitiveDateTime,
+    #[serde(with = "common_utils::custom_serde::iso8601")]
+    pub modified_at: PrimitiveDateTime,
+    #[serde(default, with = "common_utils::custom_serde::iso8601::option")]
+    pub last_synced: Option<PrimitiveDateTime>,
+    pub cancellation_reason: Option<String>,
+    pub browser_info: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/payments/payment_attempt.rs:187]: Make All Value as secret

File: crates/hyperswitch_domain_models/src/payments/payment_attempt.rs [hunk 16]
```diff
@@ -162,6 +163,111 @@ pub trait PaymentAttemptInterface {
...
+    pub routing_result: Option<serde_json::Value>,
+    pub preprocessing_step_id: Option<String>,
+    pub error_reason: Option<String>,
+    pub multiple_capture_count: Option<i16>,
+    // reference to the payment at connector side
+    pub connector_response_reference_id: Option<String>,
+    pub amount_capturable: MinorUnit,
+    pub updated_by: String,
+    pub authentication_data: Option<serde_json::Value>,
+    pub encoded_data: Option<String>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/payments/payment_attempt.rs:202]: Make this as secret

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/transformers.rs [hunk 17]
```diff
@@ -68,26 +68,6 @@ where
         .change_context(errors::ApiErrorResponse::InternalServerError)
         .attach_printable("Failed while parsing value for ConnectorAuthType")?;
 
-    let resource_id = match payment_data
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/payments/transformers.rs:71]: Yes this is intended, we do not need to populate the response beforehand.

**Review by jarnura — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Comment by Narayanbhat166** [comment]
Any changes specific to v2 will be taken up in a follow up PR.

**Review by apoorvdixit88 — APPROVED** [review]
We need to also evaluate whether we need sample data in V2

File: v2_migrations/2024-08-28-081747_recreate_ids_for_v2/up.sql [hunk 18]
```diff
@@ -30,3 +30,6 @@ ADD COLUMN IF NOT EXISTS id VARCHAR(64);
 
 ------------------------ Payment Attempt -----------------------
 ALTER TABLE payment_attempt DROP COLUMN id;
+
+ALTER TABLE payment_attempt
+ADD COLUMN IF NOT EXISTS id VARCHAR(64);
```
Comment (SanchithHegde) [inline v2_migrations/2024-08-28-081747_recreate_ids_for_v2/up.sql:35]: Are we not reverting this migration in `down.sql`?

**Review by SanchithHegde — APPROVED** [review]

**Review by prajjwalkumar17 — APPROVED** [review]

**Pull request merged**